src/net/transport.rs

```diff
@@ -30,115 +36,134 @@ impl HttpTransport {
         self.server.bind(target).await
+        */
+        todo!(); // TODO(server)
     }
 }
 
 #[async_trait]
 impl Transport for Arc<HttpTransport> {
-    type CommandStream = ReceiverStream<CommandEnvelope>;
+    type RecordsStream = Empty<Vec<u8>>; // TODO(server): resolve placeholder
 
     fn identity(&self) -> HelperIdentity {
-        self.id
+        self.identity
     }
 
-    async fn subscribe(&self, subscription: SubscriptionType) -> Self::CommandStream {
-        match subscription {
-            SubscriptionType::QueryManagement => ReceiverStream::new(
-                self.subscribe_receiver
-                    .lock()
-                    .unwrap()
-                    .take()
-                    .expect("subscribe should only be called once"),
-            ),
-            SubscriptionType::Query(query_id) => {
-                let mut ongoing_queries = self.ongoing_queries.lock().unwrap();
-                match ongoing_queries.entry(query_id) {
-                    Entry::Occupied(_) => {
-                        panic!("attempted to subscribe to commands for query id {}, but there is already a previous subscriber", query_id.as_ref())
-                    }
-                    Entry::Vacant(entry) => {
-                        let (tx, rx) = mpsc::channel(1);
-                        entry.insert(tx);
-                        ReceiverStream::new(rx)
-                    }
-                }
+    async fn send<
+        D: Stream<Item = Vec<u8>> + Send + 'static,
+        Q: QueryIdBinding,
+        S: StepBinding,
+        R: RouteParams<RouteId, Q, S>,
+    >(
+        &self,
+        dest: HelperIdentity,
+        route: R,
+        data: D,
+    ) -> Result<(), TransportError>
+    where
+        Option<QueryId>: From<Q>,
+        Option<Step>: From<S>,
+    {
+        match route.resource_identifier() {
+            RouteId::Records => {
+                let resp_future = self.clients[dest]
+                    .step(
+                        dest,
+                        <Option<QueryId>>::from(route.query_id()).expect("TODO"),
+                        <Option<Step>>::from(route.step()).as_ref().expect("TODO"),
+                        data,
+                    )
+                    .await
+                    .map_err(|_e| {
+                        io::Error::new::<String>(
+                            io::ErrorKind::ConnectionAborted,
+                            "channel closed".into(),
+                        )
+                    })?;
+                tokio::spawn(async move {
+                    resp_future
+                        .map_err(Into::into)
+                        .and_then(MpcHelperClient::resp_ok)
+                        .await
+                        .expect("failed to stream records");
+                });
+                // TODO: error handling here needs to be improved.
+                //  * We are supposed to wait until the stream has been successfully opened with the remote
```

Comment:
I was thinking "100 continue", but it looks like hyper's HTTP2 implementation doesn't support that yet. If the server sends a regular response on seeing just the request headers, is it able to change the status after seeing the body?

Reply:
I had a discussion with Alex about this and though 100 continue is a thing, I would prefer that we not use it.  For sending records, we should be able to provide an ACK in the client code that satisfies the API constraints, without waiting for the server to respond.  That means surfacing any errors later during processing instead.  Otherwise, we add a round trip unnecessarily.

Reply:
Yep, agree with Martin - acking every records request (one handshake per step) is suboptimal to say the least.

I have confirmed that for `records`/`step` API we don't really need an ack even for in-memory implementation. That leaves us with the potential dualism of `send` API - some send requests require an ack and some don't.

We could potentially solve it by adding this requirement to the `RouteParams` trait, but it is also an option to have special treatment for `step`/`records` API to return right after `POST` request is sent

I am willing to keep the acks inside the in memory implementation for now